### ISSUE TITLE
Federation Analysis: Private Key JWT Replay

### ISSUE THREAD

**Issue by selfissued** [openid/federation#94]
From: Pedram Hosseyni <[pedram.hosseyni@sec.uni-stuttgart.de](mailto:pedram.hosseyni@sec.uni-stuttgart.de)> 
Sent: Friday, September 20, 2024 2:48 AM
To: [michael_b_jones@hotmail.com](mailto:michael_b_jones@hotmail.com)
Cc: Marcus Almgren <[marcus.almgren@oidf.org](mailto:marcus.almgren@oidf.org)>; Ralf Kuesters <[ralf.kuesters@sec.uni-stuttgart.de](mailto:ralf.kuesters@sec.uni-stuttgart.de)>; Tim Würtele <[tim.wuertele@sec.uni-stuttgart.de](mailto:tim.wuertele@sec.uni-stuttgart.de)>
Subject: Federation Analysis: Private Key JWT Replay

Hi Mike,

For the Federation analysis, we found a scenario that seems to be in line with the specifications and would lead to attacks.

To put it briefly, Federation uses PAR for automatic registration, so we consider OIDC together with PAR, on top of Federation for registration. 
However, the PAR specification allows for (1) redirection endpoints that are not pre-registered and (2) using the token endpoint as the audience value of the private key JWT when authenticating at the PAR endpoint. 
However, a malicious OP could specify token endpoints of other OPs, thus, obtaining private key JWTs created by an RP that it could use at those OPs.

As the Federation specification uses PAR, we think that it should also provide enough restrictions to ensure that the resulting protocol is secure; For example, Federation already introduces restrictions on the set of authentication mechanisms that should be used for PAR.

The attached PDF contains a detailed description, including diagrams and our assumptions.

Best
Tim and Pedram

Cc: @PedramHD @SECtim

**Comment by selfissued** [comment]
[2024-09-20-private-key-JWT-attack.pdf](https://github.com/user-attachments/files/17082698/2024-09-20-private-key-JWT-attack.pdf)

**Comment by Razumain** [comment]
Just saying.

This is yet another reason why I think OpenID federation should be limited or split. I think 2 documents would be fine.

OpenID federation should only focus on the federation aspect. That is to share validated metadata and trust marks through the current structure. And that's it.

This can then be used alone or together with profiles that defines rules for registration and authentication requests.

One major reason besides this that has emerged is use of OpenID federation for the wallet infrastructure in EU with new entity types. The federation part will make sense there. The rest may not. I'm pretty sure there can be other environments that have to define their own rules as well.

**Comment by selfissued** [comment]
I agree that different environments not using OpenID Connect will have other rules.  The Federation Wallet spec https://peppelinux.github.io/federation-wallet/main.html is one example of this.

That said, we would still need to address the security disclosure in the OpenID Connect specific content whether it is in the main spec or split into a separate spec.

**Comment by peppelinux** [comment]
As discussed in our previous email conversation and agreed upon by security researchers, this vulnerability affects the entire OAuth/OpenID ecosystem, even though it was initially discovered during the OpenID Federation security analysis.

Any entity could potentially configure its metadata using another's endpoint and intercept a private_key_jwt intended for a different entity. This might happen even without federation and using dynamic client registration and or openid connect discovery.

To mitigate this risk, a recommended approach is to enforce the sanitization of endpoints, ensuring they use the issuer's HTTPS URL (entity id) as the base URL. However, as @vdzhuvinov pointed out, placing too much responsibility on the Relying Parties (RPs) is not secure. Therefore, using the entity ID (issuer HTTPS URL) in the aud claim and requiring the receiving OpenID Provider (OP) to validate it seems to be the safest solution.
